### PR TITLE
#13: feat: added a safe_cast implementation

### PR DESCRIPTION
**Pull request by kehrazy**
this adds a compile_time casting function which converts floating-point numbers to integers, avoiding undefined behaviour.

### CODE REVIEW

**Comment by kehrazy** [comment]
someone should run clang-format on this bitch, mine's brokey

**Comment by es3n1n** [comment]
So far so good, now we should replace `static_cast`s to the safe_cast where it is applicable

**Comment by kehrazy** [comment]
broke git

day = good

ot: replaced the static_cast where it's appropriate

File: portal2-internal/sdk/misc/color_t.hpp [hunk 1]
```diff
@@ -85,38 +88,44 @@ struct color_t {
...
+        this->a = static_cast<channel_t>(a * channel_limit_t::max());
         return *this;
     }
 
 public:
     // @credits: https://github.com/znsoftm/xtd/blob/master/src/xtd.drawing/src/xtd/drawing/color.cpp#L186
     [[nodiscard]] static color_t from_hsb(flt_range_t hue, flt_range_t saturation, flt_range_t brightness) {
-        if (!saturation)
-            return color_t(brightness, brightness, brightness, static_cast<channel_flt_t>(1.f));
+        if (!saturation) // @todo: kehrazy: float to bool conversion? isn't this always false in modern c++?
```
Comment (es3n1n) [inline portal2-internal/sdk/misc/color_t.hpp]: its not
https://godbolt.org/z/8vM1e4T36

Comment (kehrazy) [reply portal2-internal/sdk/misc/color_t.hpp]: so,

```cpp
if (!safe_cast<int>(saturation)) {
```
?

Comment (es3n1n) [reply portal2-internal/sdk/misc/color_t.hpp]: nah, just leave it as it is and remove comment

Comment (kehrazy) [reply portal2-internal/sdk/misc/color_t.hpp]: doned

**Review by es3n1n — APPROVED** [review]

**Pull request merged**